### PR TITLE
Add support for conversion webhook 🍸

### LINKED ISSUE #2047: Latest knative.dev/pkg adds support for conversion webhooks
This is at HEAD and will be in 0.13.

cc @vdemeester

### PR DESCRIPTION
**Pull request by vdemeester**
# Changes

This adds a new webhook controller : ConversionController.
The `HubVersion` is the stored version, and `Zygotes` stores the types
of supported versions.

The CRDs are updated to setup the conversion strategy to webhook,
pointing to the tekton-pipelines-webhook.

An update in the webhook clusterrole is required for customresourcedefinitions.

Closes #2047 

/cc @sbwsg @bobcatfish @mattmoor 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```
Add support for conversion webhook for tektoncd/pipeline CRDs
```

### CODE REVIEW

File: cmd/webhook/main.go [hunk 1]
```diff
@@ -124,6 +129,68 @@ func NewConfigValidationController(ctx context.Context, cmw configmap.Watcher) *
...
+				Zygotes: map[string]conversion.ConvertibleObject{
+					v1alpha1_: &v1alpha1.PipelineRun{},
+					v1beta1_:  &v1beta1.PipelineRun{},
+				},
+			},
+		},
+
+		// A function that infuses the context passed to ConvertTo/ConvertFrom/SetDefaults with custom metadata
+		func(ctx context.Context) context.Context {
+			return ctx
```
Comment (mattmoor) [inline cmd/webhook/main.go:189]: Did you want the ctx stuff here too?

Comment (vdemeester) [reply cmd/webhook/main.go:189]: I followed what was in `knative/serving` :stuck_out_tongue_closed_eyes:

File: cmd/webhook/main.go [hunk 2]
```diff
@@ -98,7 +103,7 @@ func NewValidationAdmissionController(ctx context.Context, cmw configmap.Watcher
 
 		// A function that infuses the context passed to Validate/SetDefaults with custom metadata.
 		func(ctx context.Context) context.Context {
-			return ctx
+			return contexts.WithUpgradeViaDefaulting(store.ToContext(ctx))
```
Comment (dprotaso) [inline cmd/webhook/main.go:106]: `contexts.WithUpgradeViaDefaulting`

Are you using this upgrade marker for anything?

Comment (vdemeester) [reply cmd/webhook/main.go:106]: yeah we do

**Comment by vdemeester** [comment]
The error is legit in this PR, see #2468

**Comment by vdemeester** [comment]
yay it is green :sun_with_face:

File: config/200-clusterrole.yaml [hunk 3]
```diff
@@ -63,6 +63,12 @@ apiVersion: rbac.authorization.k8s.io/v1
 metadata:
   name: tekton-pipelines-webhook-cluster-access
 rules:
+  # The webhook needs to be able to list and update customresourcedefinitions,
+  # mainly to update the webhook certificates.ku
+  - apiGroups: ["apiextensions.k8s.io"]
+    resources: ["customresourcedefinitions", "customresourcedefinitions/status"]
+    verbs: ["get", "list", "update", "patch", "watch"]
+    # verbs: ["get", "list", "create", "update", "delete", "patch", "watch"]
```
Comment (bobcatfish) [inline config/200-clusterrole.yaml:71]: maybe we can remove this line?

File: pkg/apis/pipeline/register.go [hunk 4]
```diff
@@ -41,3 +43,42 @@ const (
...
+	PipelineResourceResource = schema.GroupResource{
+		Group:    GroupName,
+		Resource: "pipelineresources",
+	}
+	// ConditionResource represents a Tekton Condition
+	ConditionResource = schema.GroupResource{
+		Group:    GroupName,
+		Resource: "conditions",
+	}
+)
```
Comment (bobcatfish) [inline pkg/apis/pipeline/register.go:84]: im not 100% clear on why these are required now or where they are used? does the conversion webhook automatically detect these?

File: cmd/webhook/main.go [hunk 5]
```diff
@@ -124,6 +129,68 @@ func NewConfigValidationController(ctx context.Context, cmw configmap.Watcher) *
 	)
 }
 
+func NewConversionController(ctx context.Context, cmw configmap.Watcher) *controller.Impl {
+	// nolint: golint
```
[the remark this answers is not in the document]
Comment (vdemeester) [reply cmd/webhook/main.go:133]: :sweat_smile: well, I need to find a better name than `v1beta1_` for the variable.. and I was lazy :sweat_smile:

File: config/200-clusterrole.yaml [hunk 6]
```diff
@@ -63,6 +63,12 @@ apiVersion: rbac.authorization.k8s.io/v1
 metadata:
   name: tekton-pipelines-webhook-cluster-access
 rules:
+  # The webhook needs to be able to list and update customresourcedefinitions,
+  # mainly to update the webhook certificates.ku
```
[the remark this answers is not in the document]
Comment (vdemeester) [reply config/200-clusterrole.yaml:67]: It is :man_facepalming:

File: config/200-clusterrole.yaml (hunk 3 above)
Comment (vdemeester) [reply config/200-clusterrole.yaml:71]: Indeed :sweat_smile:

File: config/300-taskrun.yaml [hunk 7]
```diff
@@ -18,6 +18,15 @@ metadata:
...
 spec:
   group: tekton.dev
+  preserveUnknownFields: false
+  validation:
+    openAPIV3Schema:
+      type: object
+      # this is a work around so we don't need to flush out the
+      # schema for each version at this time
+      #
+      # see issue: https://github.com/knative/serving/issues/912
```
[the remark this answers is not in the document]
Comment (vdemeester) [reply config/300-taskrun.yaml:28]: Right `preserveUnknownFields` and this one seems to be required to work with k8s 1.18 and the conversion webhook somehow.

[the remark this answers is not in the document]
Comment (vdemeester) [reply config/300-taskrun.yaml:28]: Updated the comments :angel:

File: pkg/apis/pipeline/register.go (hunk 4 above)
Comment (vdemeester) [reply pkg/apis/pipeline/register.go:84]: We do here : https://github.com/tektoncd/pipeline/pull/2413/files#diff-58de452513b7e8d8d3cfea23eb4ae6a8R170 (and we might do at other parts of the code). It makes it a bit easier to get `conditions.tekton.dev` or `pipelineruns.tekton.dev` in code without hard coding those.

File: cmd/webhook/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (vdemeester) [reply cmd/webhook/main.go:133]: I am not sure what variable name to use to be honest :upside_down_face:

**Comment by vdemeester** [comment]
ping @sbwsg @bobcatfish @ImJasonH this is ready for review and updated

**Pull request merged**